packages/salesforcedx-apex-debugger/src/core/breakpointService.ts

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (c) 2017, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import {
+  CliCommandExecutor,
+  CommandExecution,
+  SfdxCommandBuilder
+} from '@salesforce/salesforcedx-utils-vscode/out/src/cli';
+import {
+  ApexBreakpointLocation,
+  LineBpsInTyperef
+} from '../breakpoints/lineBreakpoint';
+import { CommandOutput } from '../utils/commandOutput';
+
+export class BreakpointService {
+  private static instance: BreakpointService;
+  private lineNumberMapping: Map<string, LineBpsInTyperef[]> = new Map();
+  private liveBreakpoints: Map<string, ApexBreakpointLocation[]> = new Map();
+
+  public static getInstance() {
+    if (!BreakpointService.instance) {
+      BreakpointService.instance = new BreakpointService();
+    }
+    return BreakpointService.instance;
+  }
+
+  public setValidLines(
+    lineNumberMapping: Map<string, LineBpsInTyperef[]>
+  ): void {
+    this.lineNumberMapping = lineNumberMapping;
+  }
+
+  public isApexDebuggerBreakpointId(id: string): boolean {
+    return id != null && id.startsWith('07b');
+  }
+
+  public getTyperefFor(uri: string, line: number): string {
+    const linesInTyperefs = this.lineNumberMapping.get(uri);
+    if (linesInTyperefs) {
+      for (const linesInTyperef of linesInTyperefs) {
+        for (const validLine of linesInTyperef.lines) {
+          if (validLine === line) {
+            return linesInTyperef.typeref;
+          }
+        }
+      }
+    }
+    return '';
```

Comment:
Should this return `string?` instead to better denote no typeref? That seems cleaner than having to check for `if (typeref && typeref.length > 0`